More container implementations

mecha opened the pull request:
Adds a variety of new container implementations:

* A prefixing decorator that allows omitting prefixes from keys
* A deprefixing decorator that requires prefixing the keys
* An aliasing decorator that can alias some or all of a container's keys
* A masking decorator that can selectively "hide" some container entries
* A callback decorator that allows fetch-time modifications to container entries
* A ~~path~~ segmenting container decorator that creates sub-segmented containers on incomplete path key access.

mecha committed:
Added masking container implementation and test

mecha committed:
Updated PHPStorm config for PHPUnit

mecha committed:
Added prefixing container implementation and test

mecha committed:
Added deprefixing container implementation and test

mecha committed:
Added missing `since` tags

mecha committed:
Added aliasing container and test

mecha committed:
Added callback container and test

mecha committed:
Added `since` tag

mecha committed:
Added path container and test

mecha committed:
Improved class PHPDoc descriptions

XedinUnknown commented on src/AliasingContainer.php:
```diff
@@ -0,0 +1,83 @@
+
+use Dhii\Container\Util\StringTranslatingTrait;
+use Psr\Container\ContainerInterface;
+use function array_key_exists;
```
Explicit use for functions could be a good idea organisation-wide.

XedinUnknown commented on src/CallbackContainer.php:
```diff
@@ -0,0 +1,70 @@
+     */
+    public function get($key)
+    {
+        return call_user_func_array($this->callback, [$this->inner->get($key), $key, $this]);
```
I'm thinking to start using `$this->callback($this->inner->get($key), $key, $this)`. AFAIK it's functionally identical, but simpler. Thoughts?

mecha replied:
I'd prefer the simpler syntax too. Need to be careful about it though as some linters or code sniffers might complain about it, since the syntax is identical to calling a method named `callback`.

XedinUnknown replied:
That's true. So, maybe store it temporarily, or try the curly braces or just parens?

mecha replied:
I found an even better way: Parentheses!

```php
return ($this->callback)(...);
```

XedinUnknown commented on src/AliasingContainer.php:
```diff
@@ -0,0 +1,83 @@
+ * A container implementation that wraps around an inner container to alias its keys, so consumers can use the aliases
+ * to fetch data from the inner container.
+ */
+class AliasingContainer implements ContainerInterface
```
I generally implement the [Dhii `ContainerInterface`][1] in Dhii packages. the implementation is a bit more complex, because the forwarding is not as simple. But it has several advantages for consumers. Do you think it's worth it?

[1]: https://github.com/Dhii/data-container-interface/blob/develop/src/ContainerInterface.php

mecha replied:
Didn't understand the latter part about the "implementation".

XedinUnknown replied:
Don't remember what I was referring to by that either. But still think that implementing the Dhii interface could be better for ISP.

mecha replied:
The Dhii interface doesn't add anything that these classes need to implement, and using the lowest-level compatible interface is natural IMO.

Not like it changes anything for consumers, but curious to hear why. If you tell me that it's just a convention that you like to stick to for Dhii packages, I'll change it.

XedinUnknown replied:
No, not a convention. Dhii interfaces add extra features that could be useful. But I'm willing to just move on for now, and return to this point in the future if it comes up.

XedinUnknown commented on src/DeprefixingContainer.php:
```diff
@@ -0,0 +1,95 @@
+     *
+     * @return string The inner key.
+     */
+    protected function getInnerKey($key)
```
Is this basically the end of the `_` prefix for protected methods, due to less likely extension?

mecha replied:
Ah! Habit. I ditched the underscore a long time ago, never really liked it anyway. If two methods have the same name but one has an underscore, to me it indicates bad pattern. Thoughts? I have no qualms adding it back in if you deem it a necessary safety measure.

XedinUnknown replied:
Well, since extension is not as likely anymore, maybe we can keep it this way for now.

XedinUnknown commented on src/MaskingContainer.php:
```diff
@@ -0,0 +1,97 @@
+    {
+        if (!$this->isExposed($key)) {
+            throw new NotFoundException(
+                $this->__('Key "%1$s" was not found in the inner container or is not exposed', [$key]),
```
IMO the message should be more explicit about the key not being exposed. Why should it mention the possibility of the key not being found?

mecha replied:
Probably a mistake from my end. The condition was likely checking for `has($key)`, but was later changed to `isExposed($key)` and the message incorrectly stayed the same.

XedinUnknown commented on src/MaskingContainer.php:
```diff
@@ -0,0 +1,97 @@
+     *                                    `false` hides the mapped key.
+     * @param bool               $default The default mask value to use for keys that are not included in the mask.
+     */
+    public function __construct(ContainerInterface $inner, array $mask, $default = true)
```
It was hard to understand how `$default` is used. Perhaps, a description of how it affects the result would help.

mecha replied:
Right! I'll add a better description.

For the record: it's a default mask, so you can set a default mask of `true` to hide everything then expose certain keys, or a default mask of `false` to expose everything and then hide certain keys.

XedinUnknown commented on src/PathContainer.php:
```diff
@@ -0,0 +1,77 @@
+     */
+    public function get($key)
+    {
+        $tKey = ltrim($key, $this->delimiter);
```
The delimiter can be more than one key. In this case, looks like the result would be not what is expected:

`ltrim($key = '-key', $delimiter = '->') // 'key'`

mecha replied:
True! Looks like the trim functions won't be good here.

XedinUnknown commented on test/functional/PathContainerTest.php:
```diff
@@ -0,0 +1,103 @@
+        $subject = $this->createSubject([$inner, $delimiter]);
+
+        $result1 = $subject->get('foo');
+        $this->assertInstanceOf('Psr\Container\ContainerInterface', $result1, 'First result in not a container');
```
Didn't quite understand what this means:

> First result in not a container

mecha replied:
The "first result" is the result of the first `get()` call in the test, since this test fetches multiple values from the container to test the `get()` method. And "not a container" mirrors the assertion: it's not a container instance.

XedinUnknown commented on src/PathContainer.php:
```diff
@@ -0,0 +1,77 @@
+ *
+ * @since [*next-version*]
+ */
+class PathContainer implements ContainerInterface
```
Had to read the test to understand what this does, so a couple of usage examples would be great!

XedinUnknown commented on src/PathContainer.php:
```diff
@@ -0,0 +1,77 @@
+ *
+ * @since [*next-version*]
+ */
+class PathContainer implements ContainerInterface
```
What are the practical benefits, i.e. realistic case scenarios of using this? I understand that this is supposed to work on inner containers which have paths for keys. In effect, doesn't this actually work as a prefixing container?

XedinUnknown replied:
Also, judging by the name of the class, I kind of expected it to instead retrieve values from a hierarchy of containers, i.e.

```
[
	'my' => [
		'path' => 'value',
	],
]
```

Running `PathContainer->get('my/path')` would return `'value'` for the above container hierarchy. I would like to have such an implementation. Are you planning to add this? And if not, when I add it, not sure what to name it then :D

mecha replied:
Yeah, needs a better name.

If a container has the key `my/special/path`, the `PathContainer` decorator would allow you to get, for instance, `my/special` (or any other non-existent key for that matter). In those cases, it creates a clone of itself that knows about that key, so when you call `path`, you finally get the value.

So you're right. It kinda does sound like a recursive prefixing container. Until an actual key is found, it will keep creating containers that are prefixed. Now I'm thinking if the concept can be generalized to some kind of generic recursive container decorator, which coupled with the prefixing decorator would do the same thing as this class.

In any case, how about `RecursivePrefixingContainer`?

XedinUnknown replied:
In my understanding, what this class does is kinda create a segment of the inner container's values, without actually retrieving that segment. And I can definitely see the use for this, given that there may be a value at `my/deep`, and the value at `my/deep/key`, where `key` is not "inside" `my/deep`. This class would circumvent that by still returning results that are consistent with configuration.

So, how about a `SegmentingContainer`?

mecha replied:
I like it.

XedinUnknown commented on src/PrefixingContainer.php:
```diff
@@ -0,0 +1,99 @@
+     * @param bool               $strict    Whether or not to fallback to un-prefixed keys if a prefixed key does not
+     *                                      exist in the inner container.
+     */
+    public function __construct(ContainerInterface $container, string $prefix, bool $strict = true)
```
IMHO it would be more intuitive and less prone to human error if the default value of `$strict` would be `false`.

mecha replied:
You'd allow access to un-prefixed keys by default?
I honestly don't have an opinion on this. I thought I did, but I started writing this reply :rofl: It sounds like a very subjective thing TBH

XedinUnknown replied:
Well, my opinion stems from the imaginary case where you say "I want a container with these keys prefixed". What would you intuitively imply about the other keys?

mecha replied:
> "I want a container with these keys prefixed" ... What would you intuitively imply about the other keys?

But this container does not allow you to choose which keys get prefixed. It's all or nothing.

XedinUnknown reviewed (changes requested):
Overall, a superb addition of useful decorators. Thank you!

mecha committed:
Shorter syntax for callback invocation

mecha committed:
Improved description of default mask

mecha committed:
Swapped order of mask and default mask ctor params

mecha committed:
Non-exposed key exception message is less vague

mecha committed:
Renamed `PathContainer` to `SegmentingContainer`

And added a detailed description with example usage.

mecha committed:
Allowing PHPUnit ^7.0 for PHP7.4 compatibility

mecha committed:
Fixed version constraint for PHPUnit

mecha committed:
Test helper now compatible with PHPUnit 7.0+

XedinUnknown commented on test/helpers/ComponentMockeryTrait.php:
```diff
@@ -263,5 +263,5 @@ protected function proxy($object): Proxy
-    abstract function getMockBuilder($className);
-}
\ No newline at end of file
+    abstract function getMockBuilder($className): MockBuilder;
```
This seems to be causing CI [builds][1] to fail.

[1]: https://travis-ci.com/Dhii/containers/jobs/279426491

mecha replied:
Indeed. Let's just remove it.

XedinUnknown commented on src/SegmentingContainer.php:
```diff
@@ -0,0 +1,110 @@
+ *
+ * @since [*next-version*]
+ */
+class SegmentingContainer implements ContainerInterface
```
I absolutely love this decorator! And the description is pretty good. But after not seeing this class for a long time, I had forgotten how it works, and couldn't immediately understand it _correctly_ from just the description. The confusing part was that this class actually only works on the key, and the internal structure is irrelevant, i.e. it is not required to be a hierarchy of containers: a flat map would work just fine, e.g.

```
[
	'config.db.host' => 'localhost',
	'config.db.port' => 3306,
]
```

However, the example in the docs suggests something else. So, I would say that what this decorator does is that it allows working with a flat map where keys are segmented as if it was a hierarchy. I.e. with the example values above, you could get `'localhost'` by doing `$c->get('config')->get('db')->get('host')`.

What do you think about adding this disambiguation to the class docs, @mecha?

XedinUnknown replied:
Do we have a container that does the opposite? As in, if you had a _container_ hierarchy (like in your original doc), and you'd have to do `$c->get('config')->get('db')->get('host')` to get the host, it would allow accessing that value by doing `$c->get('config.db.host')`. I thought there was something like that.

XedinUnknown replied:
I need one like this for objects/arrays, so as to avoid making unnecessary assumptions about their structure. Doing `isset($obj->key1->key2->key3)` is very error prone. I would like to add one such implementation that deals with objects/arrays so that `key3` can simply be checked by `$c->has('key1/key2/key3')` etc.

XedinUnknown replied:
So, #7 adds the ability to transform a data structure into a container hierarchy. Now is the time to add this container. `PathContainer` seems like a fitting name. Not sure if belongs to this PR, or #7, or a completely new one. I'd say, let's just do what's necessary to get this one merged, and then we see.

mecha replied:
Agreed. I'll update the class doc; not sure how I managed to get the example usage wrong.

Re. `PathContainer`, I think adding it to this PR is more fitting since its a container implementation. #7 is specifically related to created hierarchical structures.

XedinUnknown replied:
Fair point.

XedinUnknown commented on src/CallbackContainer.php:
```diff
@@ -0,0 +1,70 @@
+ *
+ * @since [*next-version*]
+ */
+class CallbackContainer implements ContainerInterface
```
Another small point: when I see `CallbackContainer`, I think that there will be some callback that will return values based on a key. Not that the value of the inner container will be "filtered" via the callback. How about `FilterContainer` instead?

mecha replied:
Good point.

mecha replied:
Upon further consideration, I think the term "filter" could be confusing here. Filtering typically refers to the process of reducing the source set into a subset.

The proper term for changing every element of a source set into that of another set is "mapping". So what do you think about `MappingContainer`?

XedinUnknown replied:
Sounds good to me!

mecha committed:
Renamed `CallbackContainer` to `MappingContainer`

mecha committed:
Improved `SegmentingContainer` class doc

mecha committed:
Added `HierarchicalContainer`

mecha committed:
Added `PathContainer`

mecha committed:
`SegmentingContainer` doc mentions `PathContainer`

mecha committed:
Added keys to thrown `NotFoundException` instances

mecha committed:
Fixed tests

mecha committed:
Fixed prefixing container strict mode

When in strict mode and given an unprefixed key, the prefixing container
should fail to find that key in the inner container. This was not the
case and this commit fixes that.

mecha committed:
Merge branch 'develop' into task/more-containers

XedinUnknown commented on src/HierarchyContainer.php:
```diff
@@ -0,0 +1,87 @@
+        $value = $this->data[$key];
+
+        if (is_array($value) || $value instanceof stdClass) {
+            $value = $this->data[$key] = new HierarchyContainer($value);
```
Creation of the new hierarchy container could be made more flexible by using `new static()`, instead of directly referring to this class by name.

mecha replied:
That's very bad IMO, for 2 reasons:

a) These are decorators, polymorphism should not be encouraged here
b) This class cannot possibly know the constructor signature of any of its derived classes

If you really want to support derivations of this class, we can introduce a static constructor that can be overridden.

XedinUnknown replied:
If you think that making this class more extension-friendly is a good idea, feel free to add a named constructor. Otherwise, we can leave it this way for now.

XedinUnknown commented on src/MappingContainer.php:
````diff
@@ -4,15 +4,32 @@
+ * ```php
+ * $container = new Container([
+ *      'first' => 'Paul',
+ *      'second' => 'JC',
````
Me gusta ^______^

XedinUnknown commented on src/SegmentingContainer.php:
````diff
@@ -7,42 +7,47 @@
+ * Consider the below data and a regular `$container` that provides access to it:
+ *
+ * ```php
+ * $data = [
````
Forgot to create the container from the data?

mecha replied:
I omitted it for brevity. See the sentence immediately above it:

> Consider **the below data** and a regular `$container` that provides access to it

For the example, the internal data is more important than the container that wraps around it. The doc asks the reader to consider that data, and to also keep in mind that (obviously) there is a container wrapping around it that provides untempered access to it.

XedinUnknown replied:
The reason I mentioned this is that the next example features a `SegmentingContainer` that accepts a _container_, and so it could be a little confusing as it is now. But maybe not. OK to leave it as is.

XedinUnknown commented on src/PrefixingContainer.php:
```diff
@@ -58,8 +59,12 @@ public function __construct(ContainerInterface $container, string $prefix, bool
     public function get($key)
     {
+        if (!$this->isPrefixed($key) && $this->strict) {
+            throw new NotFoundException("Key '{$key}' does not exist", 0, null, $this, $key);
```
Not a big deal, but I'd prefer `sprintf()` instead of double quotes, because it is more easily internationalized when necessary.

mecha commented on src/DeprefixingContainer.php:
```diff
@@ -0,0 +1,95 @@
+        try {
+            return $this->inner->get($this->getInnerKey($key));
+        } catch (NotFoundExceptionInterface $nfException) {
+            if ($this->strict) {
```
This check is throwing off error reporting. If the inner container fails to get a service for the prefixed key, in non-strict mode it will attempt to fetch the non-prefixed key. If that also fails, the reported exception will be with regards to the latter, even though the consumer is most likely interested in the former.

Furthermore, if the container _does_ have a service for the prefixed key but fails to fetch one of its dependencies, this error reporting will be even more vague.

Consider service `foo` depends on `bar` depends on `baz`. If the container is asked to fetch `prefix/foo` and fails to find dependency `baz`, the exception will be caught by this `try..catch` block. In non-strict mode (when the container falls back to non-prefixed keys) the container will then attempt to fetch unprefixed `foo` and report that it couldn't find `foo`, when in reality it couldn't find `baz`.

Luckily this can be mitigated easily by adding a `has` check for the unprefixed key, such that if the container doesn't have it, it won't even attempt to fetch it, regardless of strict mode. 

```php
try {
	// ...
} catch (NotFoundExceptionInterface $nfException) {
	if ($this->strict || !$this->inner->has($key)) {
		throw $nfException;
	}
}
```

This does not alter the behavior of the container - without this change the container would still throw an exception, but one that originates later from the second `get()` call. All this change does it change what exception gets thrown, which should greatly improving error reporting.

mecha committed:
Containers implement Dhii's container interface

mecha committed:
Fixed trimming for delimiters longer than 1 char

mecha committed:
Using sprintf instead of PHP template strings

mecha committed:
Throwing correct exception

mecha committed:
Fixed PHPStorm config for PHPUnit

XedinUnknown commented on src/AliasingContainer.php:
```diff
@@ -3,7 +3,8 @@
 
 use Dhii\Container\Util\StringTranslatingTrait;
-use Psr\Container\ContainerInterface;
+use Dhii\Data\Container\ContainerInterface;
```
Did I point out that this change needs to be made? 🤔 I legit don't remember. [`dhii/data-container-interface`][] is deprecated.

[`dhii/data-container-interface`]: https://github.com/Dhii/data-container-interface

mecha replied:
Here: https://github.com/Dhii/containers/pull/2#discussion_r285380423

XedinUnknown replied:
The reason it's deprecated is that many of the things from that package are now part of `dhii/collection-interface`. Some, like the additional getters for exceptions, are not IIRC. They look useful, but their utility has never actually been put to use, AFAIK.

XedinUnknown replied:
Ya, OK, thanks. Sorry about that! Could you roll back please? If you want, you can implement some of the interfaces from collections, if you think it makes stuff more useful.

mecha replied:
I'm of the opinion that the PSR container interface is enough. That's what they implemented before, and they worked. No point fixing something that's not broken.

mecha replied:
Do I remove the `dhii/data-container-interface` dependency while I'm at it? This involves removing the use of the exception interfaces from that repository.

XedinUnknown replied:
Yes, please, kind sir!

XedinUnknown reviewed (commented):
One small concern.

mecha committed:
Containers implement PSR interfaces

mecha committed:
Exceptions only implement PSR interfaces

mecha committed:
Removed `dhii/data-container-interface` dependency

XedinUnknown committed:
Merge branch 'develop' into task/more-containers

# Conflicts:
#	.idea/containers.iml
#	.idea/php-test-framework.xml
#	composer.json
#	composer.lock
#	src/CachingContainer.php
#	src/CompositeContainer.php
#	src/DelegatingContainer.php
#	src/Exception/ContainerException.php
#	src/Exception/NotFoundException.php
#	test/functional/Exception/ContainerExceptionTest.php
#	test/functional/Exception/NotFoundExceptionTest.php
#	test/helpers/ComponentMockeryTrait.php

XedinUnknown committed:
Fix redundant dep

XedinUnknown committed:
Fix problems caused by sync.

mecha committed:
Installed `dhii/collections-interface`

mecha committed:
Containers implement Dhii container interface

mecha committed:
Merge branch 'task/more-containers' of github.com:Dhii/containers into task/more-containers
